### ISSUE TITLE
Counterintuitive transformations

### ISSUE THREAD

**Issue by ArtyomLobanov** [GumTreeDiff/gumtree#103]
Hello!

I got unexpected (for me) transformations in that case:
```java
final TreeGenerator generator = new JdtTreeGenerator();
final TreeContext before = generator.generateFromFile("a.txt");
final TreeContext after = generator.generateFromFile("b.txt");
final Matcher matcher = Matchers.getInstance().getMatcher(before.getRoot(), after.getRoot());
matcher.match();
final ActionGenerator actions = new ActionGenerator(before.getRoot(), after.getRoot(), matcher.getMappings());
actions.generate().forEach(System.out::println);
```
```java
class Main {
    public static <T> void foo(Stream<? extends T> stream, Comparator<? super T> order) {
        Optional<? extends T> a = stream.min(order);
        Optional<? extends T> b = stream.max(order);
    }
}
```
```java
class Main {
    public static <T> void foo(Stream<? extends T> stream, Comparator<? super T> order) {
        List<T> elements = stream.collect(Collectors.toList());
        Optional<? extends T> a = elements.stream().min(order);
        Optional<? extends T> b = elements.stream().max(order);
    }
}
```
### Transformations:
MOV 60@@ to 8@@ at 2
INS 60@@ to 8@@ at 0
INS 74@@List<T> to 60@@ at 0
INS 59@@ to 60@@ at 1
INS 43@@List to 74@@List<T> at 0
INS 43@@T to 74@@List<T> at 1
INS 42@@elements to 59@@ at 0
INS 32@@ to 59@@ at 1
**_UPD 42@@b from b to a_**
**_UPD 42@@a from a to b_**
INS 42@@List to 43@@List at 0
INS 42@@T to 43@@T at 0
INS 42@@stream to 32@@ at 0
INS 42@@collect to 32@@ at 1
INS 32@@ to 32@@ at 2
INS 32@@ to 32@@ at 0
**_UPD 42@@max from max to min_**
INS 32@@ to 32@@ at 0
**_UPD 42@@min from min to max_**
INS 42@@Collectors to 32@@ at 0
INS 42@@toList to 32@@ at 1
INS 42@@elements to 32@@ at 0
MOV 42@@stream to 32@@ at 1
INS 42@@elements to 32@@ at 0
MOV 42@@stream to 32@@ at 1

Variables and methods names were swapped.
Maybe it's meant to be, but I thought it was best to report it because it's a bit confusing.
I use version 2.1.0 from maven repository.

Thank you!

**Comment by jrfaller** [comment]
Mmm this is odd...

Could you try with decreasing the parameters gt.bum.smt ([description](https://github.com/GumTreeDiff/gumtree/wiki/Parameters))

**Comment by jrfaller** [comment]
I have tried with the new version of the algorithm and this odd behavior is still present. I think this has to do with the subtree matcher, but I am not able yet to spot the problem :(

**Comment by jrfaller** [comment]
Hi @ArtyomLobanov !

It took me quite a long time but this is fixed by commit f45f0460b50f41f1a821e331b0325888c3da78f3 I had to introduce a new (clever, I hope :D) metric to handle this kind of case.

Cheers.
